### PR TITLE
Refactor protocol-jackson to use the new immutable protocol

### LINKED ISSUE #8837: Protocol Jackson improvements
**Description**

There are some existing issues with the current implementation of the `protocol-jackson` module. The main issues are:

1. For every type created in `protocol`, we need to add a corresponding type in `protocol-jackson`.
  1. This is error prone (as example, we already have missing types), and it's not so easy to add tests to guarantee this (though do-able)
  1. This adds unnecessary overhead for developers
1. The `Immutable` implementations are hidden - this was to reduce the API surface, but it turns out that this blocks, to some extent, interoperability with other tools and libraries, especially those operating by reflection (e.g. fixture/data generators)
1. Each of the types added in `protocol-jackson` need to also define how to deserialize nested types. This is also error prone and adds overhead for developers, and difficult to ensure via tests.
1. Copying into an immutable structure is a shallow copy, as Immutables doesn't recognize that, say, `JobRecordValue` is equivalent to `AbstractJobRecordValue`.

The proposed solution is:

- Annotate the existing types in the `protocol` module directly. These annotations are compile-time only, and as such, there is no extra dependency added to the module. 
- We combine all required annotations in a single one: `ZeebeProtocolValue` (or even just `ProtocolValue` - or suggest anything). This not only makes it easier to guarantee all types are correctly annotated, but also opens up the way for #8886 and other properties we would like to enforce in our protocol
- We would not add the Jackson bindings there, however, but instead keep `protocol-jackson` specifically to define a `ZeebeProtocolModule`, as is common with Jackson to define custom serialization.
- A new module will be created, `protocol-util`, which would add utilities to help using the protocol, notably make explicit the mapping of `ValueType`, `Intent`, and `T extends RecordValue`, and also keep track of the mapping between the interface types (e.g. `JobRecordValue`), their immutable counter part (e.g. `ImmutableJobRecordValue`), and their builders (e.g. `ImmutableJobRecordValue.Builder`). This is then used by the `protocol-jackson` module and any other module which wants this information (e.g. `protocol-test-util`).

This has several advantages:

1. Copying is now a deep copy, as you would expect
1. You can more easily use the `Immutable*` variants when you need to, and if you don't care about the serialization part, you can still use them
1. We can easily add a `ArchUnit` test which checks that the interface types are correctly annotated (which guarantees they will generate an immutable equivalent)
1. The mapping of `ValueType`, `Intent`, and values is made explicit - this can be further improved via #8886
1. It makes it very easy to add a random record generator once we have `protocol-util` and the immutable variants
1. The `protocol-jackson` module now is better defined - it's literally just a single Jackson module which you add to your object mapper and now knows how to deserialize the Zeebe protocol.
1. Having a random record factory which you can semi control (i.e. reusing `protocol-util`) makes it much easier to write exporter unit tests, where the data in the records isn't so important in most cases

See the draft PR for an idea of what it would look like.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR updates the `protocol-jackson` module to make use of the new immutable protocol classes in `protocol`, using the `protocol-util` module to discover and map the abstract and concrete types together. As mentioned in the issue, the tests are somewhat minimal, but will be fleshed out as soon as the record factory is added in the next step.

## Related issues

related to #8837 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

### CODE REVIEW

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/ZeebeProtocolModule.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import io.camunda.zeebe.protocol.util.ProtocolTypeMapping.Mapping;
+import java.util.Objects;
+
+/**
+ * A Jackson module which enables your {@link ObjectMapper} to serialize and deserialize Zeebe
+ * protocol objects, e.g. {@link Record}, {@link
+ * io.camunda.zeebe.protocol.record.value.ErrorRecordValue}, etc., essentially anything annotated
+ * with {@link io.camunda.zeebe.protocol.record.ImmutableProtocol}, out of the box.
+ *
+ * <p>To use, simply create an {@link ObjectMapper} as your normally would, and add this module.
```
Comment (korthout) [inline protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/ZeebeProtocolModule.java:24]: 🔧  typo: your -> you

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/ZeebeProtocolModule.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * <h2>Usage
```
Comment (korthout) [inline protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/ZeebeProtocolModule.java:26]: 💭 I don't think I've ever seen an H2 in javadoc before. I have no opinion on it, but I noticed the usage.

**Review by korthout — APPROVED** [review]
Nice work @npepinpe LGTM 🍸 

As I had already reviewed this, I haven't got much new to say. Please have a look at my suggestions.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/RecordMixin.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * necessary to be able to resolve the intent and the value to their correct concrete class based on
+ * the value type of the record.
+ *
+ * <p>NOTE: the type represented by {@code T} can be abstract or concrete, it doesn't matter much
+ * here.
+ *
+ * @param <T> the record value type
+ */
+@SuppressWarnings("unused")
+abstract class RecordMixin<T extends RecordValue> {
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/RecordMixin.java:29]: The annotation introspector can easily do this. The question though is how to test this :thinking: Hand-written JSON? Not ideal :sweat_smile: 

I'm not really sure I understand the one for the value however. Annotating the field is not enough, I think. You would really need to set it on the concrete class that is to be used during deserialization.

[the remark this answers is not in the document]
Comment (korthout) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/RecordMixin.java:29]: >Annotating the field is not enough, I think.

I remember testing this locally, with an annotation on the getValue method. And that works as long as you're deserializing an entire record. If you want to only deserialize a record value then you'll need to have annotations on the concrete classes AFAIK.

[the remark this answers is not in the document]
Comment (korthout) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/RecordMixin.java:29]: >Hand-written JSON?

That's what I did 😅  You could also use jackson to modify the json AFAIK

[the remark this answers is not in the document]
Comment (npepinpe) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/RecordMixin.java:29]: I added some things, let me know what you think :+1:

**Review by korthout — APPROVED** [review]
👏  LGTM

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**